### PR TITLE
kie-issues#237: Implement DMN 1.4 Boxed iterator expression

### LINKED ISSUE #237: Implement DMN 1.4 Boxed `iterator` expression
We should provide enhancement for the boxed expression editor so user do not need to type manually FEEL expressions that iterate collelction:
`for [variable] in [collection] return [expression]`

You can find more details in [**10.2.1.10 Boxed iterator**](https://www.omg.org/spec/DMN/1.4/PDF/)

Screenshot 2023-04-25 094324

```[tasklist]
### Acceptance criteria
- [ ] new boxed expression is available in context menu invoked on empty cell
- [ ] new boxed expression has two columns, the left column is readonly
```

### LINKED ISSUE #238: Implement DMN 1.4 `every/some -> in -> satisfies` boxed expression
We should provide enhancement of existing boxed expression editor, so user no not need to type manually FEEL expression with form:
`every/some [variable] in [collection] satisfies [condition]`

instead of it, boxed expression editor should offer box with three rows

You can find more details in [**10.2.1.10 Boxed iterator**](https://www.omg.org/spec/DMN/1.4/PDF/)

Screenshot 2023-04-25 094613

```[tasklist]
### Acceptance criteria
- [ ] new boxed expression is available in context menu invoked on empty cell
- [ ] new boxed expression has two columns, the left column is readonly
```

### PR DESCRIPTION
**Pull request by danielzhe**
Closes: https://github.com/apache/incubator-kie-issues/issues/237
Closes: https://github.com/apache/incubator-kie-issues/issues/238

Both tickets were made together because the `iterator expression` covers both as it is in the spec. See [**10.2.1.10 Boxed iterator**](https://www.omg.org/spec/DMN/1.4/PDF/)

### CODE REVIEW

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        } else if (args.rowIndex === 1) {
+          // Do not inline this variable for type safety. See https://github.com/microsoft/TypeScript/issues/241
+          const ret: BoxedIterator = {
+            ...prev,
+            in: { expression: undefined! }, // SPEC DISCREPANCY
+          };
+          return ret;
+        } else if (args.rowIndex === 2) {
+          if (prev.__$$element === "for") {
+            const ret: BoxedFor = {
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx:294]: Missing comment "// Do not inline this variable for type safety. See https://github.com/microsoft/TypeScript/issues/241"

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            };
+            return ret;
+          } else {
+            // Do not inline this variable for type safety. See https://github.com/microsoft/TypeScript/issues/241
+            const iterator: BoxedIterator = {
+              ...prev,
+              satisfies: { expression: undefined! }, // SPEC DISCREPANCY
+            };
+            return iterator;
+          }
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx]: Defaulting to "every" and "some", while I think we should actively check for those types and throw an Error if anything else is found.

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        } else {
+          throw new Error("ConditionalExpression shouldn't have more than 3 rows.");
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx]: Wrong error message.

File: packages/boxed-expression-component/src/resizing/WidthMaths.ts [hunk 4]
```diff
@@ -328,6 +328,34 @@ export function getExpressionResizingWidth(
...
+    return (
+      resizingWidth ??
+      CONDITIONAL_EXPRESSION_LABEL_COLUMN_WIDTH +
+        Math.max(
+          CONDITIONAL_EXPRESSION_CLAUSE_COLUMN_MIN_WIDTH,
+          ...nestedExpressions.map((e) => getExpressionResizingWidth(e, resizingWidths, widthsById))
+        ) +
+        CONDITIONAL_EXPRESSION_EXTRA_WIDTH
+    );
+  }
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/resizing/WidthMaths.ts]: Not using the ITERATOR_EXPRESSION constants.

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,320 @@
...
+          );
+        } else {
+          return (
+            <IteratorExpressionCell
+              iteratorClause={props.data[props.rowIndex]}
+              {...props}
+              parentElementId={parentElementId}
+            />
+          );
+        }
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx]: Defaulting to this, while I think we should explicitly check for rowIndex being 1 and 2, as iterator expressions shouldn't have more than 3 rows.

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+  const getIterableRowLabel = useCallback(
+    (rowNumber: number) => {
+      if (rowNumber === 0) {
+        return expression.__$$element;
+      } else if (rowNumber === 1) {
+        return "in";
+      } else {
+        if (expression.__$$element === "for") {
+          return "return";
+        } else return "satisfies";
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx]: Same here. Defaulting to "satisfies", while we should check for explicit cases we support and throw an error for all other unexpected cases.

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        group: i18n.terms.selection.toUpperCase(),
+        items: [{ name: i18n.terms.copy, type: BeeTableOperation.SelectionCopy }],
+      },
+    ];
+  }, [i18n]);
+
+  const getIterableRowLabel = useCallback(
+    (rowNumber: number) => {
+      if (rowNumber === 0) {
+        return expression.__$$element;
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx]: Using `__$$element` here is not ideal, as it is a coincidence that those two things have the same value. Please explicitly return the name of the expression based on what type of iterator it is.

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionCell.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            }
+        }
+      });
+    },
+    [rowIndex, setExpression]
+  );
+
+  const currentExpression = useMemo(() => {
+    if (typeof iteratorClause.child !== "string") {
+      return iteratorClause.child?.expression as any;
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionCell.tsx]: Why "as any"?

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionCell.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                ...prev,
+                return: {
+                  expression: getNewExpression(),
+                },
+              };
+            } else {
+              return {
+                ...prev,
+                satisfies: {
+                  expression: getNewExpression(),
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionCell.tsx]: Why not pass `prev` here, like you're doing at the `in` part?

File: packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+            >
+              <InlineEditableTextInput
+                {...props}
+                value={props.data[props.rowIndex].child as string}
+                onChange={(updatedValue) => {
+                  setExpression((prev: BoxedIterator) => {
+                    return {
+                      ...prev,
+                      "@_iteratorVariable": updatedValue,
+                    };
```
Comment (tiagobento) [inline packages/boxed-expression-component/src/expressions/IteratorExpression/IteratorExpressionComponent.tsx]: Please follow the same strategy of extracting the return value to a variable, explicitly typing it and adding the comment saying why this is necessary.

**Review by tiagobento — CHANGES_REQUESTED** [review]
@danielzhe Thanks for the PR! I left some comments inline.

**Review by tiagobento — APPROVED** [review]

**Review by ljmotta — APPROVED** [review]
@danielzhe Amazing new features. The code is looking good.

**Comment by jaimeizuz** [comment]
Hi @ljmotta  @tiagobento @danielzhe 

I've just tested this new functionality and I'm getting the following exception:

`{
  "name": "Error",
  "message": "IteratorExpression can't have more than 3 rows.",
  "stack": "Error: IteratorExpression can't have more than 3 rows.
    at child (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4951260)
    at aa (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4331465)
    at Gs (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4383558)
    at Dl (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4370837)
    at Sl (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4370765)
    at kl (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4370628)
    at _l (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4367615)
    at https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4317387
    at t.unstable_runWithPriority (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4533910)
    at Yr (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4317164)
    at Kr (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4317332)
    at $r (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4317267)
    at Ie (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4388077)
    at Qt (https://business-modeller-app.eu.ngrok.io/new-dmn-editor-envelope.js:2:4295958)"
}`

https://github.com/apache/incubator-kie-tools/assets/138981666/cdd63bdb-4bb4-4387-93a7-e06c2b977283

**Build Version:**
0.0.0 (daily-dev) @ 236931cf7a3299ac81d1fd84c7228cb1033496ae (2024-04-18 08:31:17)
**Kogito Runtimes Version:**
999-20240414-SNAPSHOT
**Quarkus Version:**
3.2.10.Final
**Dev deployments Base image URL:**
quay.io/kie-tools/dev-deployment-base-image:daily-dev
**Dev deployments Kogito Quarkus Blank App image URL:**
quay.io/kie-tools/dev-deployment-kogito-quarkus-blank-app-image:daily-dev
**Dev deployments DMN Form webapp image URL:**
quay.io/kie-tools/dev-deployment-dmn-form-webapp-image:daily-dev
**Extended Services version:**
0.0.0
**Commit SHA:**
236931cf7a

Google Chrome Version 123.0.6312.123 (Official Build) (64-bit) in Windows 10

**Comment by danielzhe** [comment]
@jaimeizuz Ty for your report. It got broken in the code review, I'm already fixing it.

**Pull request merged**